### ISSUE TITLE
How can we mock a protocol in an dependency framework?

### ISSUE THREAD

**Issue by jibeex** [MakeAWishFoundation/SwiftyMocky#212]
Swifty works perfectly in my project. But I cannot find a way to mock some service(protocol) that is located in a framework. Is there a way to do that correctly?

**Comment by amichnia** [comment]
Hi @jibeex, thank for the info.

I will think how to approach that. If framework has accessible sources (most of pods), you can add a path to the file in mockfile. Let's say the protocol in a framework is called `ProtocolA`.

Unfortunately you probably can't make it conform to "AutoMockable", so I see 2 options here:

1. Create your own protocol like that `protocol ProtocolB: ProtocolA, AutoMockable {}` and use it instead if that's possible.

2. Extend `ProtocolA` in your code, adding "comment annotation" like: 
```
//sourcery: AutoMockable
extension ProtocolA {}
```

Let me know if any of that worked. I will try to find more structural approach for that.

**Comment by amichnia** [comment]
Just to let you know - we are working on an Adapters module, which might solve this issue, but it is under deep development now, with no real release date yet.

**Comment by jibeex** [comment]
@amichnia I have tried your methods this week. Both methods that you proposed are not working properly. A mock is created, but methods of the protocol from the framework are not populated.

**Comment by amichnia** [comment]
@jibeex I'm testing it right now. If I can make it work, I'll prepare example in our sample project/documentation. If this is more work, I'll have to postpone it until version 4. Please stay tuned

**Comment by jibeex** [comment]
@amichnia Sounds great. Thanks.

**Comment by amichnia** [comment]
@jibeex I made some tests. At the moment (SwiftyMocky 3.3.4) the only way to solve it is to create "dummy" protocol based on interface of the protocol you want to mock.

You can do it two ways:

1. When you have no access to the sources, create `Ext` protocol, and copy methods from the protocol you want to mock. Example for `URLSessionDelegate`:

```swift
//sourcery: AutoMockable
//sourcery: ObjcProtocol
@objc protocol URLSessionDelegateExt: URLSessionDelegate {

    @objc @available(OSX 10.9, *)
    optional func urlSession(_ session: URLSession, didBecomeInvalidWithError error: Error?)

    @objc @available(OSX 10.9, *)
    optional func urlSession(_ session: URLSession, didReceive challenge: URLAuthenticationChallenge, completionHandler: @escaping (URLSession.AuthChallengeDisposition, URLCredential?) -> Void)
}
```

That would allow to generate `URLSessionDelegateExtMock`, which you can effectively use as mock of `URLSessionDelegate`.

2. When you have access to the sources:

This is basically same, but if you add path to original protocol source file, you can omit copying methods definitions. Mock should inherit methods from known types (known means Sourcery has access to the source files)

It would look somewhat like:

```swift
//sourcery: AutoMockable
protocol ExternalProtocol: Alamofire.SessionDelegate {}
```

```yml
# Mockfile
    sources:
     - ...
     - ./Pods/Alamofire/SessionDelegate.swift # Add path to sources
```
